### PR TITLE
Update MLflow python client to enable tracking on Windows

### PR DESCRIPTION
**Pull request by tomasatdatabricks**
## What changes are proposed in this pull request?
Various changes to the Mlflow python client to make the tracking and artifacts work on Windows.

#### 1. Removed path module on artifact repository. 
All artifacts-repository paths are now expected to be posix paths. This is so that it is clear which path notation to use when referencing mlflow artifacts and to ensure platform independence. 

#### 2. All artifact path operations done  with posixpath module. All local file path operations done with local os.path module. 
The use has been mixed and e.g. artifact uploads / downloads would not work properly on windows.

#### 3. Default artifact root for file store is now a file uri (file:///path).
This is necessary because we can not tell whether absolute windows path is local path or a uri with one letter scheme (unlikely but e.g. "c:path" is a valid uri with scheme == "c".). 
 
## How is this patch tested?
Standard set of tests + small python tests on windows machine. All the small tests with the exception of projects ran successfully. In particular, the following command returns no errors on windows (with python3):

```
pytest --cov=mlflow --verbose --ignore=tests/h2o --ignore=tests/keras --ignore=tests/pytorch --ignore=tests/pyfunc --ignore=tests/sagemaker --ignore=tests/sklearn  --ignore=tests/spark --ignore=tests/tensorflow --ignore tests/azureml --ignore tests/projects tests

============================== warnings summary ===============================
tests/store/test_sftp_artifact_repo.py::test_artifact_uri_factory
  c:\programdata\anaconda3\lib\site-packages\pysftp\__init__.py:61: UserWarning: Failed to load HostKeys from C:\Users\Administrator\.ssh\known_hosts.  You will need to explicitly load HostKeys (cnopts.hostkeys.load(filename)) or disableHostKey checking (cnopts.hostkeys = None).
    warnings.warn(wmsg, UserWarning)

-- Docs: http://doc.pytest.org/en/latest/warnings.html
============================= 5 tests deselected ==============================
====== 459 passed, 2 skipped, 5 deselected, 1 warnings in 189.50 seconds ======

```
 
## Release Notes
Update MLflow python client to enable tracking on Windows. Windows users can now leverage MLflow python client and cli to create, track and inspect experiments as well as log and download artifacts. 
 
### Is this a user-facing change? 

- [ ] No. Add the `rn/none` label, then you can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
  
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [x] Tracking
- [ ] Projects 
- [x] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ X] Python

### Please add one label to the PR so it can be classified correctly in the release notes. Options:
 
* `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
* `rn/none` - No description will be included. The PR will be mentioned just by the PR number in the "Small Bugfixes and Documentation Updates" section
* `rn/feature` - A new user-facing feature worth mentioning in the release notes
* `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
* `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/tracking/test_rest_tracking.py [hunk 1]
```diff
@@ -51,9 +52,14 @@ def _await_server_down_or_die(process, timeout=60):
     """Waits until the local flask server process is terminated."""
     print('Awaiting termination of server process...')
     start_time = time.time()
-    while process.is_alive() and time.time() - start_time < timeout:
+
+    def wait():
```
Comment (sueann) [inline tests/tracking/test_rest_tracking.py]: why was this change required? is it more platform independent?

File: tests/tracking/test_rest_tracking.py [hunk 2]
```diff
@@ -68,11 +74,20 @@ def _init_server(backend_uri, root_artifact_uri):
...
     }
     with mock.patch.dict(os.environ, env):
-        process = Process(target=lambda: app.run(LOCALHOST, server_port))
-        process.start()
+        cmd = ["python",
+               "-c",
+               'from mlflow.server import app; app.run("{hostname}", {port})'.format(
+                   hostname=LOCALHOST, port=server_port)]
+        process = Popen(cmd)
+        # def get_output():
```
Comment (sueann) [inline tests/tracking/test_rest_tracking.py]: remove?

File: tests/tracking/test_rest_tracking.py [hunk 3]
```diff
@@ -220,12 +251,12 @@ def test_create_run_all_args(mlflow_client, parent_run_id_kwarg):
         },
         "parent_run_id": parent_run_id_kwarg
     }
+    experiment_name = 'Run A Lot (parent_run_id=%s)' % (parent_run_id_kwarg)
```
Comment (sueann) [inline tests/tracking/test_rest_tracking.py]: unused?

File: mlflow/tracking/artifact_utils.py [hunk 4]
```diff
@@ -57,10 +53,8 @@ def _download_artifact_from_uri(artifact_uri, output_path=None):
     :param output_path: The local filesystem path to which to download the artifact. If unspecified,
                         a local output path will be created.
     """
-    artifact_path_module = \
-        get_artifact_repository(artifact_uri).get_path_module()
-    artifact_src_dir = artifact_path_module.dirname(artifact_uri)
-    artifact_src_relative_path = artifact_path_module.basename(artifact_uri)
+    artifact_src_dir = posixpath.dirname(artifact_uri)
```
Comment (sueann) [inline mlflow/tracking/artifact_utils.py]: this would be a problem if artifact_uri is a file path on windows right? this function directly powers functions like `load_model` which i assumed would take in windows paths.

File: mlflow/store/file_store.py [hunk 5]
```diff
@@ -172,8 +177,9 @@ def list_experiments(self, view_type=ViewType.ACTIVE_ONLY):
 
     def _create_experiment_with_id(self, name, experiment_id, artifact_uri):
         self._check_root_dir()
-        meta_dir = mkdir(self.root_directory, experiment_id)
-        artifact_uri = artifact_uri or build_path(self.artifact_root_uri, experiment_id)
+        meta_dir = mkdir(self.root_directory, str(experiment_id))
+        artifact_uri = artifact_uri or path_to_local_file_uri(
```
Comment (sueann) [inline mlflow/store/file_store.py]: is this where you said if a user supplies the artifact_uri that is local we don't covert to `file:/...`? i guess the only thing i see that will be weird is `get_artifact_uri` will return a weirdly-formatted uri like `c:\user\given\artifact\uri/relative/path` which should be ok...

**Review by sueann — COMMENTED** [review]
thanks for doing this! still need to review local_artifact_repo but wanted to send over some questions / comments before i can get back to it in a little bit.

File: tests/projects/test_projects.py [hunk 6]
```diff
@@ -199,15 +199,15 @@ def test_run_local_git_repo(local_git_repo,
     submitted_run.wait()
     validate_exit_status(submitted_run.get_status(), RunStatus.FINISHED)
     # Validate run contents in the FileStore
-    run_id = submitted_run.run_id
+    run_uuid = submitted_run.run_id
```
Comment (sueann) [inline tests/projects/test_projects.py]: i think these are all meant to be `run_id` from aaron's PR?

File: tests/projects/utils.py [hunk 7]
```diff
@@ -46,7 +48,9 @@ def build_docker_example_base_image():
 @pytest.fixture()
 def tracking_uri_mock(tmpdir):
     try:
-        mlflow.set_tracking_uri(os.path.join(tmpdir.strpath, 'mlruns'))
+        path = os.path.abspath(os.path.join(tmpdir.strpath, 'mlruns'))
```
Comment (sueann) [inline tests/projects/utils.py]: did you mean to undo these projects changes, or do they not matter anyway?

File: mlflow/utils/file_utils.py [hunk 8]
```diff
@@ -328,12 +329,10 @@ def _copy_file_or_tree(src, dst, dst_dir=None):
...
         dst_subpath = os.path.join(dst_dir, dst_subpath)
     dst_path = os.path.join(dst, dst_subpath)
-
-    dst_dirpath = os.path.dirname(dst_path)
-    if not os.path.exists(dst_dirpath):
-        os.makedirs(dst_dirpath)
-
     if os.path.isfile(src):
+        dst_dirpath = os.path.dirname(dst_path)
+        if not os.path.exists(dst_dirpath):
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/utils/file_utils.py]: This was just moved down into if-branch. Previously this was done even if we were copying the whole tree which is unnecessary and shutils on windows don't like it. It would fail with path already exists.

File: mlflow/store/artifact_repo.py [hunk 9]
```diff
@@ -79,12 +71,14 @@ def download_artifacts(self, artifact_path, dst_path=None):
...
         :return: Absolute path of the local filesystem location containing the downloaded artifacts.
         """
+
         # TODO: Probably need to add a more efficient method to stream just a single artifact
         # without downloading it, or to get a pre-signed URL for cloud storage.
 
         def download_artifacts_into(artifact_path, dest_dir):
-            basename = self.get_path_module().basename(artifact_path)
-            local_path = build_path(dest_dir, basename)
+            print("download artifacts into", artifact_path, dest_dir)
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/store/artifact_repo.py]: debugging. I will remove it after all tests are green.I guess I should put back the [WIP] to the title.

File: mlflow/store/artifact_repo.py [hunk 10]
```diff
@@ -37,7 +29,7 @@ def log_artifact(self, local_file, artifact_path=None):
 
         :param local_file: Path to artifact to log
         :param artifact_path: Directory within the run's artifact directory in which to log the
-                              artifact
+                              artifact.
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/store/artifact_repo.py]: That's what I was thinking. But I guess it would not hurt to mention it everywhere.

File: mlflow/store/dbfs_artifact_repo.py [hunk 11]
```diff
@@ -65,15 +66,11 @@ def _get_dbfs_path(self, artifact_path):
     def _get_dbfs_endpoint(self, artifact_path):
         return "/dbfs%s" % self._get_dbfs_path(artifact_path)
 
-    def get_path_module(self):
-        import posixpath
-        return posixpath
-
     def log_artifact(self, local_file, artifact_path=None):
-        basename = self.get_path_module().basename(local_file)
+        basename = os.path.basename(local_file)
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/store/dbfs_artifact_repo.py]: Yes, I believe it was.

File: mlflow/store/dbfs_artifact_repo.py [hunk 12]
```diff
@@ -92,10 +89,10 @@ def log_artifacts(self, local_dir, artifact_path=None):
         for (dirpath, _, filenames) in os.walk(local_dir):
             artifact_subdir = artifact_path
             if dirpath != local_dir:
-                rel_path = self.get_path_module().relpath(dirpath, local_dir)
-                artifact_subdir = self.get_path_module().join(artifact_path, rel_path)
+                rel_path = os.path.relpath(dirpath, local_dir)
+                artifact_subdir = posixpath.join(artifact_path, rel_path)
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/store/dbfs_artifact_repo.py]: oh good catch, I'll update this. 

posixpath.normpath does not work unfortunately as it does not recognize '\' as separator. I added a method for this in file_utils, I'll call it here.

File: mlflow/store/file_store.py (hunk 5 above)
Comment (tomasatdatabricks) [reply mlflow/store/file_store.py]: Yes, we do not convert user supplied artifact uri. 

Windows absolute path would not work though as it would look like a uri with single character scheme. For absolute paths on windows, user must use the file:///<drive>:/<path> pattern. 

In case of relative path with '\' separator, we do not convert it and mixed separators should work in windows.

File: mlflow/store/ftp_artifact_repo.py [hunk 13]
```diff
@@ -61,53 +62,50 @@ def _size(self, full_file_path):
...
-            artifact_dir = self.get_path_module().join(self.path, artifact_path) \
+            artifact_dir = posixpath.join(self.path, artifact_path) \
                 if artifact_path else self.path
             self._mkdir(artifact_dir)
             with open(local_file, 'rb') as f:
                 ftp.cwd(artifact_dir)
-                ftp.storbinary('STOR ' + self.get_path_module().basename(local_file), f)
+                ftp.storbinary('STOR ' + os.path.basename(local_file), f)
 
     def log_artifacts(self, local_dir, artifact_path=None):
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/store/ftp_artifact_repo.py]: I agree this could be centralized similarly to download artifacts.

File: mlflow/store/ftp_artifact_repo.py [hunk 14]
```diff
@@ -61,53 +62,50 @@ def _size(self, full_file_path):
...
 
-        local_dir = self.get_path_module().abspath(local_dir)
+        local_dir = os.path.abspath(local_dir)
         for (root, _, filenames) in os.walk(local_dir):
             upload_path = dest_path
             if root != local_dir:
-                rel_path = self.get_path_module().relpath(root, local_dir)
-                upload_path = self.get_path_module().join(dest_path_re, rel_path)
+                rel_path = os.path.relpath(root, local_dir)
+                upload_path = posixpath.join(dest_path_re, rel_path)
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/store/ftp_artifact_repo.py]: Good catch again. Will fix.

File: mlflow/store/gcs_artifact_repo.py [hunk 15]
```diff
@@ -51,24 +48,24 @@ def log_artifact(self, local_file, artifact_path=None):
...
 
-        local_dir = self.get_path_module().abspath(local_dir)
+        local_dir = os.path.abspath(local_dir)
         for (root, _, filenames) in os.walk(local_dir):
             upload_path = dest_path
             if root != local_dir:
-                rel_path = self.get_path_module().relpath(local_dir, root)
-                upload_path = self.get_path_module().join(dest_path, rel_path)
+                rel_path = os.path.relpath(local_dir, root)
+                upload_path = posixpath.join(dest_path, rel_path)
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/store/gcs_artifact_repo.py]: Will fix. Thanks!

File: mlflow/store/s3_artifact_repo.py [hunk 16]
```diff
@@ -27,41 +28,37 @@ def _get_s3_client(self):
...
         s3_client = self._get_s3_client()
-        local_dir = self.get_path_module().abspath(local_dir)
+        local_dir = os.path.abspath(local_dir)
         for (root, _, filenames) in os.walk(local_dir):
             upload_path = dest_path
             if root != local_dir:
-                rel_path = self.get_path_module().relpath(root, local_dir)
-                upload_path = self.get_path_module().join(dest_path, rel_path)
+                rel_path = os.path.relpath(root, local_dir)
+                upload_path = posixpath.join(dest_path, rel_path)
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/store/s3_artifact_repo.py]: Will fix

File: mlflow/tracking/artifact_utils.py (hunk 4 above)
Comment (tomasatdatabricks) [reply mlflow/tracking/artifact_utils.py]: Yes but I don't think we can make this work with windows paths, at least not without some sacrifices - i.e. what if the artifact name contains '\'? 

I think it's ok not to accept windows paths here because the argument is artifact path which we say is always posix. User can still supply posix path on windows or fully qualified uri.

File: mlflow/tracking/utils.py [hunk 17]
```diff
@@ -69,7 +69,7 @@ def get_tracking_uri():
     elif env.get_env(_TRACKING_URI_ENV_VAR) is not None:
         return env.get_env(_TRACKING_URI_ENV_VAR)
     else:
-        return os.path.abspath(DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH)
+        return path_to_local_file_uri(os.path.abspath(DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH))
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/tracking/utils.py]: Not that I know of as long as all of our code handles file:// uris correctly. Hopefully our tests would catch these.

File: mlflow/tracking/utils.py [hunk 18]
```diff
@@ -89,8 +89,7 @@ def _is_databricks_uri(uri):
 
 
 def _get_file_store(store_uri, **_):
-    path = urllib.parse.urlparse(store_uri).path if store_uri else None
-    return FileStore(path, path)
+    return FileStore(file_utils.local_file_uri_to_path(store_uri), store_uri)
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/tracking/utils.py]: I don't think we do. 
This is an internal method and we are not calling it with None and I don't think we should.

File: tests/projects/test_projects.py (hunk 6 above)
Comment (tomasatdatabricks) [reply tests/projects/test_projects.py]: Good catch! Broken merge. I'll fix these.

Comment (tomasatdatabricks) [reply tests/projects/test_projects.py]: Actually I wanted to revert this whole file since I decided projects were out of scope of this PR..

File: tests/projects/utils.py (hunk 7 above)
Comment (tomasatdatabricks) [reply tests/projects/utils.py]: I do want to revert these, good catch.

File: tests/tracking/test_rest_tracking.py (hunk 2 above)
Comment (tomasatdatabricks) [reply tests/tracking/test_rest_tracking.py]: Yup, will do!

File: tests/tracking/test_rest_tracking.py (hunk 1 above)
Comment (tomasatdatabricks) [reply tests/tracking/test_rest_tracking.py]: The multiprocessing on windows is weird. I did not find a way how to start a Flask server using the Process abstraction. I had to use Popen instead so the process object is different. unfortunately python2 does not have a timeout argument in wait() and I did not find other way of fetching the return code other then calling wait(). And since we want to avoid a hang, the wait is called from a new thread.

File: tests/tracking/test_rest_tracking.py (hunk 3 above)
Comment (tomasatdatabricks) [reply tests/tracking/test_rest_tracking.py]: True, will remove!

File: tests/utils/test_validation.py [hunk 19]
```diff
@@ -4,15 +4,17 @@
 from mlflow.exceptions import MlflowException
 from mlflow.entities import Metric, Param, RunTag
 from mlflow.protos.databricks_pb2 import ErrorCode, INVALID_PARAMETER_VALUE
+
```
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply tests/utils/test_validation.py]: Yeah I think I can revert this.

File: mlflow/store/local_artifact_repo.py [hunk 20]
```diff
@@ -1,54 +1,71 @@
...
 
     def _download_file(self, remote_file_path, local_path):
-        shutil.copyfile(
-            self.get_path_module().join(self.artifact_uri, remote_file_path), local_path)
-
-    def _resolve_path(self, artifact_path):
-        artifact_dir = self.get_path_module().join(self.artifact_uri, artifact_path) \
-            if artifact_path else self.artifact_uri
-        return artifact_dir
+        print("_download_file", remote_file_path, local_path)
```
Comment (sueann) [inline mlflow/store/local_artifact_repo.py]: Remove debugging prints in this fn. otherwise the file lgtm. Thanks!

File: mlflow/tracking/utils.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (sueann) [reply mlflow/tracking/utils.py]: I wonder why wr had the if statement before?

**Review by sueann — COMMENTED** [review]
looks like we're almost there! remaining are just the one debugging print statement and the `_get_file_store` `None` question. and i guess more tests need to pass :-P

File: mlflow/tracking/utils.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/tracking/utils.py]: Good question. It looks like FileStore actually accepts uri=None in which case it sues default artifact root directory. 
However, neither Client nor Fluent api would ever call _get_file_store with None for store uri, the both use default uri in that case. And as far as I know, there are no other callers.

I think it makes sense to have only one source of default value - default tracking uri filled in by client or fluent api. 
So, in conclusion, I think we should not allow store_uri=None here and we should remove the None handling logic from FileStore as well.

**Review by sueann — APPROVED** [review]
lgtm - thanks for adding travis!
